x/programs/cmd/simulator/cmd/request.go

```diff
@@ -249,6 +222,11 @@ func (c *runCmd) runStepFunc(
 		// TODO: implement readonly for now just don't charge for gas
 		result, _, err := programExecuteFunc(ctx, c.log, db, testContext, params[1:], method, math.MaxUint64)
 		output := resultToOutput(result, err)
+		// check if the db has any new changes
+		if db.HasChanges() {
+			changes := db.Changes()
+			return fmt.Errorf("%w: %s, %s", ErrInvalidRequest, "readonly request should not have any changes", changes)
+		}
```

Comment:
I think that the `ReadOnly` endpoint should also not be able to deploy, send value, and call with value. Maybe it would make more sense to build these restrictions as part of the VM as compared to the simulator (in host calls). You could call `programExecuteFunc` function with an additional `readOnly bool` flag possibly